### ISSUE TITLE
Splitting out the list of providers in a standalone JSON file

### ISSUE THREAD

**Issue by SylvainCorlay** [leaflet-extras/leaflet-providers#457]
Currently, the available providers are listed in a JavaScript object that is [defined inline in the source code of leaflet-providers](https://github.com/leaflet-extras/leaflet-providers/blob/1.13.0/leaflet-providers.js#L78-L1093).

It would be really useful to make that information available directly in a JSON file that could be used from other projects. For example, [xyzservices](https://github.com/geopandas/xyzservices/blob/main/provider_sources/_parse_leaflet_providers.py) extracts that information from leaflet-provider.js to make the same information available to Python. Xyzservices has some adoption in the scientific Python ecosystem ([Geopandas](https://github.com/geopandas/geopandas), [ipyleaflet](https://github.com/jupyter-widgets/ipyleaflet), [contextily](https://github.com/geopandas/contextily), [leafmap](https://github.com/giswqs/leafmap), [bokeh](https://github.com/bokeh/bokeh), and others) - meaning that this project has in fact a lot of indirect adoption...

This separate JSON file could even be published independently of the leaflet-providers JS source, which would probably make it valuable for other users.

**Comment by martinfleis** [comment]
Hi @brunob, we already have that magic in `xyzservices`. We scrape your JS to get that JSON but the result is slightly adapted to work within our use case. But the Python code for that is here https://github.com/geopandas/xyzservices/blob/main/provider_sources/_parse_leaflet_providers.py and a resulting JSON here https://github.com/geopandas/xyzservices/blob/main/provider_sources/leaflet-providers-parsed.json.

**Comment by brunob** [comment]
@martinfleis do you think we can use your script on our repo with a github ci or anything like that ?

**Comment by martinfleis** [comment]
@brunob Yeah, we do that within this GHA https://github.com/geopandas/xyzservices/blob/main/.github/workflows/update_providers.yaml using this [makefile](https://github.com/geopandas/xyzservices/blob/main/Makefile)

**Comment by brunob** [comment]
Nice, let's hope that @jieter could look into it or propose an alternative method :)

**Comment by SylvainCorlay** [comment]
We could generate this JSON as a one-off, include the JSON in your repository, and then read the JSON from your JavaScript instead of having the object hard coded in the code base.

This would make it a single source of truth.

**Comment by jieter** [comment]
Yes, it would be nice to maintain the list in a machine readable format, and generate `leaflet-providers.js` from it + a template. I am not sure about JSON though. In the current definitions, we use some comments: ([1](https://github.com/leaflet-extras/leaflet-providers/blob/master/leaflet-providers.js#L487-L492) [2](https://github.com/leaflet-extras/leaflet-providers/blob/master/leaflet-providers.js#L606-L614) [3](https://github.com/leaflet-extras/leaflet-providers/blob/master/leaflet-providers.js#L911-L913) [4](https://github.com/leaflet-extras/leaflet-providers/blob/master/leaflet-providers.js#L941-L942)), which I think is a valuable capability not possible in classic json. 

Other options would be [JSONC](https://github.com/Microsoft/node-jsonc-parser), but that would require consumers of the file to use non-standard parsers too.

Any suggestions?

**Comment by martinfleis** [comment]
What about using YAML instead if we need comments? That feels as a bit more standard solution than JSONC.

**Comment by SylvainCorlay** [comment]
The big advantage of JSON is that you can just import it in your JS code an assign it to the variable you are setting in the current version.

**Comment by brunob** [comment]
+1 for JSON too

**Comment by martinfleis** [comment]
One option to keep comments in the JSON without using non-standard parsers would be to include an optional `note` or `comment` field with a string that is currently in the comment and ignore it while parsing the JSON to show it on your website or within xyzservices.

**Comment by rafaqz** [comment]
We would also like this at [TileProviders.jl](https://github.com/JuliaGeo/TileProviders.jl) (which is used in Leaflet.jl but is generalised so everyone can use it)

Currently we pull from geopandas xyzservices because its easier to parse

To minimise downstream workload I suggest that rather than YAML or a new JSON format we use the JSON format xyzservices has already defined.

**Comment by vincerubinetti** [comment]
This would be great, and looking at the code I can't imagine it would be too hard to split it out. I'd be happy to contribute a PR, but it looks like there's a lot of old unmerged PRs.

I've switched from Leaflet to Open Layers for a variety of technical reasons, but there doesn't seem to be an equivalent plugin like this one for OL.

Currently it looks like the geopandas extraction of the JSON here is more up to date, so I will graciously take that and import it into my project 😇

**Comment by vincerubinetti** [comment]
Also, here's a TypeScript schema for the json data structure.

```ts
type Provider = {
    url?:              string;
    max_zoom?:         number;
    html_attribution?: string;
    attribution?:      string;
    name?:             string;
    bounds?:           Array<number[]>;
    variant?:          string;
    apikey?:           string;
    min_zoom?:         number;
    ext?:              string;
    accessToken?:      string;
    tileSize?:         number;
    zoomOffset?:       number;
    id?:               string;
    key?:              string;
    subdomains?:       string;
    style?:            string;
    apiKey?:           string;
    opacity?:          number;
    mapID?:            string;
    app_id?:           string;
    app_code?:         string;
    base?:             string;
    type?:             string;
    language?:         string;
    format?:           string;
    size?:             string;
    time?:             string;
    tilematrixset?:    string;
    tms?:              boolean;
    detectRetina?:     boolean;
    apiVersion?:       string;
    subscriptionKey?:  string;
    timeStamp?:        string;
}

type Providers = Record<string, Provider | Record<string, Provider>>;

```

<details><summary>snippet i used to generate the type</summary>

Run this in a browser dev console:

```js
providers = {}; // paste json here
provider = {};
for (const [service, variants] of Object.entries(providers)) {
  if (service === "_meta") continue;
  if ("url" in variants) provider = { ...provider, ...variants };
  else
    for (const variant of Object.values(variants))
      provider = { ...provider, ...variant };
}
```

`provider` now contains all the possible key/values

Then use this service to generate the type from the object literal:
https://app.quicktype.io/?l=ts

</details>
